packages/effect/src/Effect.ts

```diff
@@ -5198,3 +5197,53 @@ export const fromNullable: <A>(value: A) => Effect<NonNullable<A>, Cause.NoSuchE
 export const optionFromOptional: <A, E, R>(
   self: Effect<A, E, R>
 ) => Effect<Option.Option<A>, Exclude<E, Cause.NoSuchElementException>, R> = effect.optionFromOptional
+
+/**
+ * @since 2.0.0
+ * @category constructors
+ */
+export const AccessTag: <const Id extends string>(id: Id) => <Self, Shape>() =>
+  & Context.TagClass<Self, Id, Shape>
+  & {
+    $: {
+      [k in keyof Shape]: Shape[k] extends Effect<infer A, infer E, infer R> ? Effect<A, E, Self | R>
+        : Effect<Shape[k], never, Self>
+    }
+  }
+  & {
+    [k in keyof Shape as Shape[k] extends (...args: Array<any>) => any ? k : never]: Shape[k] extends
+      (...args: infer Args) => Effect<infer A, infer E, infer R> ? (...args: Args) => Effect<A, E, Self | R>
+      : Shape[k] extends (...args: infer Args) => infer A ? (...args: Args) => Effect<A, never, Self>
+      : never
+  } = (id) => () => {
+    const limit = Error.stackTraceLimit
+    Error.stackTraceLimit = 2
+    const creationError = new Error()
+    Error.stackTraceLimit = limit
+    function TagClass() {}
+    Object.setPrototypeOf(TagClass, TagProto)
+    TagClass.key = id
+    Object.defineProperty(TagClass, "stack", {
+      get() {
+        return creationError.stack
+      }
+    })
+    // @ts-ignore
+    TagClass["$"] = new Proxy({}, {
+      get(_target: any, prop: any, _receiver) {
+        // @ts-expect-error
+        return core.andThen(TagClass, (s) => s[prop])
```

Comment:
maybe cache?

Reply:
caching proxy leads to Max Call Stack

Reply:
I mean like this https://github.com/Effect-TS/effect/pull/2230/files

Reply:
I think this would be slower, not sure, the issue caching would try to prevent is proxy access but proxy is accessed anyway, for the rest a cache lookup or a single function call have comparable overhead

Reply:
my point was more like garbage, creating a new function instead of lookup and return old. at least for the function case. but sounds like micro optimisations, especially considering all the functions we create and destroy on our side of the pond :D